File: go/04_trees_and_graphs/03_list_of_depth.go
package trees_and_graphs

import (
	"container/list"
	. "github.com/holyketzer/ctci_v6/lib"
)

// n - size of array
// Time = O(n) Mem = O(n)
func ListsOfDepthA(tree *BinaryTree) map[int]([]*BTNode) {
	res := map[int]([]*BTNode){}

	queue := list.New()
	queue.PushBack(tree.Root)
	queue.PushBack(0)

	for queue.Len() > 0 {
		e := queue.Front()
		curr := e.Value.(*BTNode)
		queue.Remove(e)

		e = queue.Front()
		level := e.Value.(int)
		queue.Remove(e)

		if res[level] == nil {
			res[level] = []*BTNode{}
		}
		res[level] = append(res[level], curr)

		if curr.Left != nil {
			queue.PushBack(curr.Left)
			queue.PushBack(level + 1)
		}

		if curr.Right != nil {
			queue.PushBack(curr.Right)
			queue.PushBack(level + 1)
		}
	}

	return res
}

File: ruby/spec/04_trees_and_graphs/03_list_of_depth.rb
require_relative '02_minimal_tree'

# n - size of array
# Time = O(n) Mem = O(n)
def lists_of_depth_a(tree)
  res = []
  q = [tree.root, 0]

  while !q.empty? do
    curr = q.shift
    level = q.shift

    res[level] ||= LinkedList.new
    res[level].push(curr)

    if curr.left
      q << curr.left
      q << level + 1
    end

    if curr.right
      q << curr.right
      q << level + 1
    end
  end

  res
end

RSpec.describe 'lists_of_depth' do
  include MinimalTree

  %i(a).each do |implementation|
    describe "#{implementation} case" do
      subject { send("lists_of_depth_#{implementation}", tree).map(&:to_array).map { |arr| arr.map(&:value) } }

      context 'perfect tree' do
        let(:tree) { array_to_bst_a([1, 2, 3, 4, 5, 6, 7]) }

        it do
          expect(subject[0]).to eq [4]
          expect(subject[1]).to eq [2, 6]
          expect(subject[2]).to eq [1, 3, 5, 7]
        end
      end

      context 'not perfect tree' do
        let(:tree) { array_to_bst_a([1, 2, 3, 4, 5, 6, 7, 8]) }

        it do
          expect(subject[0]).to eq [5]
          expect(subject[1]).to eq [3, 7]
          expect(subject[2]).to eq [2, 4, 6, 8]
          expect(subject[3]).to eq [1]
        end
      end
    end
  end
end
